modules/json/json-parser.c

```diff
@@ -160,24 +161,33 @@ json_parser_process_single(struct json_object *jso,
       break;
     case json_type_array:
     {
-      gint i, plen;
-
-      g_string_assign(key, obj_key);
-
-      plen = key->len;
+      g_string_truncate(value, 0);
 
-      for (i = 0; i < json_object_array_length(jso); i++)
+      type = LM_VT_LIST;
+      for (gint i = 0; i < json_object_array_length(jso); i++)
         {
-          g_string_truncate(key, plen);
-          g_string_append_printf(key, "[%d]", i);
-          json_parser_process_single(json_object_array_get_idx(jso, i),
-                                     prefix,
-                                     key->str, msg);
+          struct json_object *el = json_object_array_get_idx(jso, i);
+          GString *element_value = scratch_buffers_alloc();
+          LogMessageValueType element_type;
+
+          if (json_parser_extract_value(el, element_value, &element_type))
+            {
+              if (i != 0)
+                g_string_append_c(value, ',');
+              str_repr_encode_append(value, element_value->str, element_value->len, NULL);
+            }
+          else
+            {
+              /* unknown type, encode the entire array as JSON */
+              g_string_assign(value, json_object_to_json_string_ext(jso, JSON_C_TO_STRING_PLAIN));
+              type = LM_VT_JSON;
+              break;
+            }
         }
-      break;
-    }
+
```

Comment:
This resolves #1073 :)

I could not find this documented anywhere, I think we can change this without providing a compat layer.

Reply:
This is actually really exciting because $(format-json) as well supports lists `LM_VT_LIST`, so we produce almost identical results when applying json-parser and then format-json (except for embedded objects).

Reply:
It even works with embedded objects, as in that case json-parser() recognizes there are non-trivial-typed elements in the array, so we fall back to using LM_VT_JSON for the entire array, which will be inserted into the JSON verbatim.

Indeed it does resolve #1073.